### PR TITLE
Refactor `gear-core-processor` crate and cover by tests

### PR DESCRIPTION
**Pull request by shamilsan**
- Part of #1496
- [x] Add tests
- [x] Add test for pages to be updated

### CODE REVIEW

File: core-processor/src/executor.rs [hunk 1]
```diff
@@ -266,40 +275,45 @@ fn prepare_memory<A: ProcessorExt, M: Memory>(
...
 fn get_pages_to_be_updated<A: ProcessorExt>(
-    mut old_pages_data: BTreeMap<PageNumber, PageBuf>,
+    old_pages_data: BTreeMap<PageNumber, PageBuf>,
     new_pages_data: BTreeMap<PageNumber, PageBuf>,
     static_pages: WasmPageNumber,
 ) -> BTreeMap<PageNumber, PageBuf> {
+    if A::LAZY_PAGES_ENABLED {
+        // In lazy pages mode we update some page data in storage,
+        // when it has been write accessed, so no need to compare old and new page data.
+        // FIXME: I'd add a `cfg` attr here, as it is not cheap
```
Comment (shamilsan) [inline core-processor/src/executor.rs:285]: Need to decide how we deal with log traces.

Comment (shamilsan) [reply core-processor/src/executor.rs:281]: Added `verbose` feature for now.

File: core-processor/src/processor.rs [hunk 2]
```diff
@@ -438,11 +439,10 @@ pub fn process<
             }
         },
         Err(e) => match e.reason {
-            ExecutionErrorReason::InitialMemoryBlockGasExceeded
-            | ExecutionErrorReason::GrowMemoryBlockGasExceeded
-            | ExecutionErrorReason::LoadMemoryBlockGasExceeded => {
-                process_allowance_exceed(dispatch, program_id, e.gas_amount.burned())
-            }
+            // FIXME: What about ProgramData?
```
Comment (breathx) [inline core-processor/src/processor.rs:442]: What's wrong with program data?

File: core-processor/src/processor.rs [hunk 3]
```diff
@@ -311,12 +311,13 @@ pub fn prepare(
         Err(reason) => {
             log::debug!("Failed to charge for module instantiation or memory pages: {reason:?}");
             return match reason {
-                ExecutionErrorReason::InitialMemoryBlockGasExceeded
-                | ExecutionErrorReason::GrowMemoryBlockGasExceeded
-                | ExecutionErrorReason::LoadMemoryBlockGasExceeded
-                | ExecutionErrorReason::ModuleInstantiationBlockGasExceeded => {
-                    prepare_allowance_exceed(dispatch, program_id, gas_counter)
-                }
+                // FIXME: What about ProgramData?
```
Comment (breathx) [inline core-processor/src/processor.rs:314]: Same here

File: core/src/program.rs [hunk 4]
```diff
@@ -93,13 +93,13 @@ impl Program {
...
     /// Get reference to memory pages.
-    pub fn get_allocations(&self) -> &BTreeSet<WasmPageNumber> {
+    pub fn allocations(&self) -> &BTreeSet<WasmPageNumber> {
         &self.allocations
     }
 
-    /// Get mut reference to memory pages.
-    pub fn get_allocations_mut(&mut self) -> &mut BTreeSet<WasmPageNumber> {
-        &mut self.allocations
+    /// Set reference to memory pages.
```
Comment (breathx) [inline core/src/program.rs:100]: For this fn and one above `fn allocations` incorrect doc comment

File: core-processor/Cargo.toml [hunk 5]
```diff
@@ -11,13 +11,10 @@ gear-core-errors = { path = "../core-errors", features = ["codec"] }
...
 log = { version = "0.4.17", default-features = false }
-anyhow = { version = "1.0.65", default-features = false }
 derive_more = "0.99.17"
 
-[dev-dependencies]
-hex = "0.4.3"
-
 [features]
 strict = []
+verbose = []
```
Comment (breathx) [inline core-processor/Cargo.toml:20]: Seems that superfluous work for now, so I'd remove it

Comment (breathx) [reply core-processor/Cargo.toml:20]: Otherwise need to adapt all scripts etc

File: core-processor/src/executor.rs [hunk 6]
```diff
@@ -132,18 +147,13 @@ pub(crate) fn charge_gas_for_instantiation(
     gas_allowance_counter: &mut GasAllowanceCounter,
 ) -> Result<(), ExecutionErrorReason> {
     let amount = gas_per_byte * code_length as u64;
-
     log::trace!("Charge {} for module instantiation", amount);
```
Comment (breathx) [inline core-processor/src/executor.rs:150]: such trace may be also included in `charge_gas` fn using debug formatter for gas operation

File: core-processor/src/executor.rs [hunk 7]
```diff
@@ -488,3 +499,148 @@ pub fn execute_wasm<
         allocations: info.allocations,
     })
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (breathx) [inline core-processor/src/executor.rs:504]: I'd also add test for pages to be updated with and without lazy pages feature

**Review by breathx — APPROVED** [review]

File: core-processor/src/processor.rs (hunk 2 above)
Comment (shamilsan) [reply core-processor/src/processor.rs]: Old comment, removed.

File: core-processor/src/processor.rs (hunk 3 above)
Comment (shamilsan) [reply core-processor/src/processor.rs]: Removed too.

File: core/src/program.rs (hunk 4 above)
Comment (shamilsan) [reply core/src/program.rs]: Renamed.

File: core-processor/Cargo.toml (hunk 5 above)
Comment (shamilsan) [reply core-processor/Cargo.toml]: Ok, I remove it at the moment but need to add it in the future as `log::trace` forces to iterate `new_pages_data` in `executor::get_pages_to_be_updated`.

File: core-processor/src/executor.rs (hunk 6 above)
Comment (shamilsan) [reply core-processor/src/executor.rs]: Super idea! Done.

File: core-processor/src/executor.rs (hunk 7 above)
Comment (shamilsan) [reply core-processor/src/executor.rs:501]: Added them.

**Pull request merged**